File: triangle/recursive_with_memoization.rb
class Triangle
  
  def initialize(file)
    @file = file
  end
  
  def get_array(file)
    text = File.open(file, 'r').read
    text.each_line.map do |line|
      line.strip().split(' ').map do |n|
        { :num => n.to_i }
      end
    end
  end
  
  def get_root(array)
    array.each_with_index do |row, i|
      row.each_with_index do |node, j|
        next_row = array[i+1]
        unless next_row.nil?
          node[:left] = next_row[j]
          node[:right] = next_row[j+1]
        end
      end
    end
    array[0][0]
  end
  
  def walk_tree(node)
    return 0 if node.nil?
    return node[:memo_sum] unless node[:memo_sum].nil?
    node[:memo_sum] = node[:num] + [walk_tree(node[:left]), walk_tree(node[:right])].max
  end
  
  def get_max
    walk_tree( get_root( get_array( @file)))
  end
end

puts Triangle.new('triangle.txt').get_max

File: triangle/recursive_with_memoization.js
var fs = require('fs');

var max = function (file) {
  
  var text = fs.readFileSync(file, 'ascii');
  var array = text.split('\n').map(function(row) {
    return row.trim().split(' ').map(function(word) {
      return { num: parseInt(word) };
    });
  });
  
  array.forEach(function(row, i) {
    row.forEach(function(node, j) {
      var next_row = array[i+1];
      if (next_row) {
        node.left = next_row[j];
        node.right = next_row[j+1];
      };
    });
  });
  var root = array[0][0];
  
  return (function walk_tree(node) {
    if (!node) return 0;
    if (typeof node.memo_sum !== 'undefined') return node.memo_sum;
    node.memo_sum = node.num + Math.max(walk_tree(node.left), walk_tree(node.right));
    return node.memo_sum;
  })(root);
  
}  

console.log(max('triangle.txt'))